lib/fd.c

```diff
@@ -0,0 +1,42 @@
+// SPDX-FileCopyrightText: 2024, Skyler Ferrante <sjf5462@rit.edu>
+// SPDX-License-Identifier: BSD-3-Clause
+
+/**
+ * To protect against file descriptor omission attacks, we open the std file
+ * descriptors with /dev/null if they are not already open. Code is based on
+ * fix_fds from sudo.c.
+ */
+
+#include "prototypes.h"
+
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
```

Comment:
I would sort(1) these.

Anyway, LGTM.

```
Reviewed-by: Alejandro Colomar <alx@kernel.org>
```

Reply:
Oh, and I don't think we need <sys/types.h> nor <sys/stat.h>, right?  And we need <stdlib.h> for abort(3).

Reply:
At this point, maybe I should add
```
Co-developed-by: Alejandro Colomar <alx@kernel.org>
Signed-off-by: Alejandro Colomar <alx@kernel.org>
```

:)

Reply:
> Oh, and I don't think we need <sys/types.h> nor <sys/stat.h>, right? And we need <stdlib.h> for abort(3).

Hmm, I thought we did for open.

```

Reply:
> At this point, maybe I should add
> 
> ```
> Co-developed-by: Alejandro Colomar <alx@kernel.org>
> Signed-off-by: Alejandro Colomar <alx@kernel.org>
> ```
> 
> :)

> Co-authored-by: Alejandro Colomar <alx@kernel.org>

:)

Reply:
> > Oh, and I don't think we need <sys/types.h> nor <sys/stat.h>, right? And we need <stdlib.h> for abort(3).
> 
> Hmm, I thought we did for open.
> 
> ```
> NAME
>        open, openat, creat - open and possibly create a file
> 
> SYNOPSIS
>        #include <sys/types.h>
>        #include <sys/stat.h>
>        #include <fcntl.h>
> ```
> 
> I can edit it if there not actually needed.
```
open(2)                   System Calls Manual                  open(2)

NAME
     open, openat, creat - open and possibly create a file

LIBRARY
     Standard C library (libc, -lc)

SYNOPSIS
     #include <fcntl.h>

     int open(const char *pathname, int flags, ...
                /* mode_t mode */ );
```